File: libb/lib/user/ml.dart
import 'package:flutter/material.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:ml_algo/ml_algo.dart';
import 'package:ml_dataframe/ml_dataframe.dart';


class FirebaseService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  Future<List<Map<String, dynamic>>> fetchMaintenanceData() async {
    var result = await _firestore.collection('maintenance').get();
    return result.docs.map((doc) => doc.data()).toList();
  }

  Future<List<Map<String, dynamic>>> fetchFuelData() async {
    var result = await _firestore.collection('fuel').get();
    return result.docs.map((doc) => doc.data()).toList();
  }

  Future<List<Map<String, dynamic>>> fetchTripData() async {
    var result = await _firestore.collection('trips').get();
    return result.docs.map((doc) => doc.data()).toList();
  }

  Future<List<Map<String, dynamic>>> fetchRouteData() async {
    var result = await _firestore.collection('routes').get();
    return result.docs.map((doc) => doc.data()).toList();
  }
}

class MaintenanceModel {
  late DataFrame _dataFrame;
  late LinearRegressor _model;

  Future<void> trainModel(List<Map<String, dynamic>> data) async {
    _dataFrame = DataFrame(data as Iterable<Iterable>);

    final targetName = 'maintenance_cost';
    _model = LinearRegressor(_dataFrame, targetName);
  }

  double predict(List<double> input) {
    final inputDataFrame = DataFrame([
      ['amount', 'fueltype', 'cost', 'routeid'],
      input,
    ]);
    final prediction = _model.predict(inputDataFrame);
    return prediction.rows.first.first;
  }
}

class MyAppp extends StatefulWidget {
  @override
  _MyApppState createState() => _MyApppState();
}

class _MyApppState extends State<MyAppp> {
  final FirebaseService _firebaseService = FirebaseService();
  final MaintenanceModel _maintenanceModel = MaintenanceModel();

  @override
  void initState() {
    super.initState();
    _fetchAndTrainModel();
  }

  Future<void> _fetchAndTrainModel() async {
    var maintenanceData = await _firebaseService.fetchMaintenanceData();
    var fuelData = await _firebaseService.fetchFuelData();
    var tripData = await _firebaseService.fetchTripData();
    var routeData = await _firebaseService.fetchRouteData();

    // Combine and preprocess data here
    var combinedData = maintenanceData.map((maintenance) {
      var fuel = fuelData.firstWhere((f) => f['vehicleid'] == maintenance['vehicleid']);
      var trip = tripData.firstWhere((t) => t['vehicleid'] == maintenance['vehicleid']);
      var route = routeData.firstWhere((r) => r['vehicleid'] == maintenance['vehicleid']);

      return {
        'vehicleid': maintenance['vehicleid'],
        'amount': fuel['amount'],
        'fueltype': fuel['fueltype'],
        'cost': maintenance['cost'],
        'routeid': trip['routeid'],
        'maintenance_cost': maintenance['cost'],
        'routeid': route['id'],
      };
    }).toList();

    await _maintenanceModel.trainModel(combinedData);
  }

  Future<void> _predictMaintenance() async {
    var input = ['amount', 'fueltype', 'cost', 'routeid'];
    var prediction = _maintenanceModel.predict(input.cast<double>());
    print('Prediction: $prediction');
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Predictive Analytics'),
        ),
        body: Center(
          child: ElevatedButton(
            onPressed: _predictMaintenance,
            child: Text('Predictive Analytics'),
          ),
        ),
      ),
    );
  }
}

File: libb/lib/user/ml.py
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
import joblib

# Assuming data is fetched and converted to DataFrame
maintenance_data = pd.read_csv('maintenance.dart')
fuel_data = pd.read_csv('fuel.dart')
trip_data = pd.read_csv('trip.dart')
routes_data = pd.read_csv('routes.dart')

# Merge dataframes on vehicleid
merged_data = maintenance_data.merge(fuel_data, on='vehicleid')
merged_data = merged_data.merge(trip_data, on='vehicleid')
merged_data = merged_data.merge(routes_data, on='vehicleid')

# Feature selection and preprocessing
features = merged_data[['amount', 'fueltype', 'cost', 'routeid']]
labels = merged_data['maintenance_cost']

X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2)
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train the model
model = RandomForestRegressor()
model.fit(X_train_scaled, y_train)

# Save the model
joblib.dump(model, 'maintenance_model.pkl')
